In [ ]:
# ============================================================
# STAGE 1 - CONTAINER 1
# DATASET OVERVIEW
# ============================================================

print("DATASET OVERVIEW")
print("=" * 50)
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

df = pd.read_csv("/content/Dataset(1).csv")

print("Dataset loaded successfully!")
print("Shape:", df.shape)
rows, columns = df.shape

print(f"Number of records : {rows:,}")
print(f"Number of columns : {columns}")

print("\nDataset memory usage:")
memory_mb = df.memory_usage(deep=True).sum() / (1024 ** 2)
print(f"{memory_mb:.2f} MB")

print("\nIndex information:")
print(f"Index type        : {type(df.index).__name__}")
print(f"Index start       : {df.index.min()}")
print(f"Index end         : {df.index.max()}")

DATASET OVERVIEW


Dataset loaded successfully!
Shape: (250000, 27)
Number of records : 250,000
Number of columns : 27

Dataset memory usage:


344.52 MB

Index information:
Index type        : RangeIndex
Index start       : 0
Index end         : 249999


In [ ]:
# ============================================================
# STAGE 1 - CONTAINER 2
# COLUMN AND DATA-TYPE PROFILE
# ============================================================

print("COLUMN AND DATA-TYPE PROFILE")
print("=" * 70)

profile = pd.DataFrame({
    "Column": df.columns,
    "Data Type": df.dtypes.astype(str).values,
    "Non-Null Count": df.notna().sum().values,
    "Unique Values": df.nunique(dropna=True).values
})

display(profile)

COLUMN AND DATA-TYPE PROFILE


,Column,Data Type,Non-Null Count,Unique Values
0,Transaction ID,object,250000,250000
1,time_stamp,object,245357,189975
2,Transaction_Type,object,244939,55
3,merchant category,object,244936,91
4,amount_INR,object,245364,19230
5,Transaction Status,object,244959,35
6,sender_agegroup,object,245437,58
7,receiver_age_group,object,245428,57
8,Sender State,object,245428,91
9,sender bank,object,245434,81


In [ ]:
# ============================================================
# STAGE 1 - CONTAINER 3
# RECORD INSPECTION
# ============================================================

print("FIRST 5 RECORDS")
print("=" * 50)
display(df.head())

print("\nLAST 5 RECORDS")
print("=" * 50)
display(df.tail())

print("\n5 RANDOM RECORDS")
print("=" * 50)
display(df.sample(5, random_state=42))

FIRST 5 RECORDS


,Transaction ID,time_stamp,Transaction_Type,merchant category,amount_INR,Transaction Status,sender_agegroup,receiver_age_group,Sender State,sender bank,Receiver_Bank,device type,Network Type,Fraud Flag,hour of day,Day Of Week,is weekend,battery level %,App Version,device storage free gb,screen brightness %,Device Language,GPS accuracy m,nearby wifi count,session duration sec,notifications last hour,merchant distance km
0,TXN0000000001,10/8/2024 15:17,P2P,Entertainment,868,SUCCESS,26-35,18-25,Delhi,Axis,SBI,Android,4G,0,15,Tuesday,0,28,3.4.2,245.3,14,Hindi,7.1,4,98,3,2.64
1,TXN0000000002,2024/04/11 06:56:00,P2M,Grocery,1011,SUCCESS,26-35,26-35,Uttar Pradesh,ICICI,Axis,iOS,4G,0,6,Thursday,0,69,3.3.0,187.9,87,Marathi,22,6,24,8,-5
2,TXN0000000003,4/2/2024 13:27,P2P,Grocery,477,SUCCESS,26-35,36-45,Karnataka,Yes Bank,PNB,Android,4G,0,13,Tuesday,-1,13,3.4.2,113.2,51,Hindi,7.6,8,99,6,1.45
3,TXN0000000004,1/7/2024 10:09,P2P,Fuel,2784,SUCCESS,26-35,26-35,Delhi,ICICI,PNB,Android,5G,0,10,Sunday,1,25,3.3.0,172.1,98,Hindi,22.8,3,151,8,0.26
4,TXN0000000005,1/23/2024 19:04,P2P,Shopping,990,SUCCESS,26-35,18-25,Delhi,Axis,Yes Bank,iOS,WiFi,0,19,Tuesday,0,35,4.0.1,224.3,17,English,6.1,4,40,4,NaN



LAST 5 RECORDS


,Transaction ID,time_stamp,Transaction_Type,merchant category,amount_INR,Transaction Status,sender_agegroup,receiver_age_group,Sender State,sender bank,Receiver_Bank,device type,Network Type,Fraud Flag,hour of day,Day Of Week,is weekend,battery level %,App Version,device storage free gb,screen brightness %,Device Language,GPS accuracy m,nearby wifi count,session duration sec,notifications last hour,merchant distance km
249995,TXN0000249996,11/8/2024 22:41,Recharge,Food,373,SUCCESS,36-45,26-35,Telangana,ICICI,HDFC,iOS,5G,0,22,Friday,0,39,3.3.0,213.2,56,English,6,2,31,7,0.75
249996,TXN0000249997,12/15/2024 2:58,P2P,Utilities,2025,SUCCESS,36-45,26-35,Rajasthan,HDFC,Yes Bank,Android,5G,0,2,Sunday,1,91,NaN,134.8,18,Hindi,15.5,5,51,4,2.42
249997,TXN0000249998,11/27/2024 16:33,P2P,Food,468,SUCCESS,26-35,18-25,West Bengal,ICICI,ICICI,Android,4G,0,16,Wednesday,0,36,3.1.0,192.4,NaN,English,99999,7,51,4,unknown
249998,TXN0000249999,1/5/2024 13:31,Recharge,Healthcare,284,SUCCESS,18-25,36-45,Andhra Pradesh,Axis,ICICI,iOS,4G,0,13,Friday,0,90,3.5.0,33.7,61,NaN,12.5,4,121,3,7.53
249999,TXN0000250000,1/17/2024 15:23,P2P,Entertainment,531,SUCCESS,18-25,26-35,Telangana,Axis,Yes Bank,Android,4G,0,15,Wednesday,0,14,3.5.0,79.4,47,English,8.1,3,207,3,2.96



5 RANDOM RECORDS


,Transaction ID,time_stamp,Transaction_Type,merchant category,amount_INR,Transaction Status,sender_agegroup,receiver_age_group,Sender State,sender bank,Receiver_Bank,device type,Network Type,Fraud Flag,hour of day,Day Of Week,is weekend,battery level %,App Version,device storage free gb,screen brightness %,Device Language,GPS accuracy m,nearby wifi count,session duration sec,notifications last hour,merchant distance km
38683,TXN0000038684,5/12/2024 15:12,P2M,Healthcare,988,SUCCESS,36-45,26-35,Karnataka,SBI,SBI,Android,4G,0,15,Fridayy,1,58,3.4.2,153.1,75,Hindi,unknown,6,20,4,2.79
64939,TXN0000064940,1/23/2024 11:35,P2P,Transport,57,SUCCESS,18-25,26-35,Delhi,Yes Bank,IndusInd,iOS,WiFi,0,11,Tuesday,no,55,3.4.2,213.5,49,English,7.1,0,40,7,6.74
3954,TXN0000003955,7/15/2024 5:05,NaN,Shopping,5872,FAILED,18-25,26-35,Uttar Pradesh,Axis,PNB,Android,4G,0,5,mondayy,0,36,3.3.0,71,40,Marathi,7,4,49,3,4.17
120374,TXN0000120375,12/26/2024 16:42,P2M,Transport,134,SUCCESS,26-35,36-45,Rajasthan,HDFC,ICICI,iOS,5G,0,unknown,Thursday,0,57,3.5.0,114.5,-10,Telugu,16.5,7,127,9,0.8
172861,TXN0000172862,8/25/2024 12:19,P2P,Entertainment,325,SUCCESS,18-25,26-35,Tamil Nadu,SBI,ICICI,Android,4G,0,25,Sunday,1,73,3.4.2,147.5,60,Hindi,29.7,-1,73,4,2.82


In [ ]:
# ============================================================
# STAGE 1 - CONTAINER 4
# DATA QUALITY INSPECTION
# ============================================================

print("DATA QUALITY INSPECTION")
print("=" * 60)

# Missing values
missing = df.isnull().sum()

# Duplicate rows
duplicates = df.duplicated().sum()

# Total missing cells
total_missing = missing.sum()

print(f"Total missing cells : {total_missing:,}")
print(f"Duplicate rows      : {duplicates:,}")

print("\nMissing values by column:")
missing_table = pd.DataFrame({
    "Column": missing.index,
    "Missing Values": missing.values
})

display(missing_table)

# Explicit conclusion
print("\nDATA QUALITY RESULT")
print("=" * 60)

if total_missing == 0:
    print("✓ No missing values were found.")

if duplicates == 0:
    print("✓ No duplicate rows were found.")

DATA QUALITY INSPECTION


Total missing cells : 124,747
Duplicate rows      : 0

Missing values by column:


,Column,Missing Values
0,Transaction ID,0
1,time_stamp,4643
2,Transaction_Type,5061
3,merchant category,5064
4,amount_INR,4636
5,Transaction Status,5041
6,sender_agegroup,4563
7,receiver_age_group,4572
8,Sender State,4572
9,sender bank,4566



DATA QUALITY RESULT
✓ No duplicate rows were found.


In [ ]:
# ============================================================
# STAGE 1 - CONTAINER 5
# INITIAL NUMERICAL PROFILE
# ============================================================

print("INITIAL NUMERICAL PROFILE")
print("=" * 70)

# The current dataset contains numerical fields stored as text/object
# values because several fields contain blank entries.  For Stage 1,
# we only profile these fields; the original dataframe is not modified.

numeric_data = pd.DataFrame(index=df.index)

for column in df.columns:
    converted = pd.to_numeric(df[column], errors="coerce")

    if converted.notna().any():
        numeric_data[column] = converted

numeric_columns = numeric_data.columns.tolist()

print(f"Number of numerical columns identified: {len(numeric_columns)}")

print("\nNumerical columns:")
for i, column in enumerate(numeric_columns, start=1):
    print(f"{i}. {column}")

print("\nBasic numerical summary:")

if len(numeric_columns) > 0:
    display(numeric_data[numeric_columns].describe().T)
else:
    print("No numerical columns were identified for profiling.")

print("\nNumerical profiling note:")
print("Numeric-looking fields were converted only in a temporary profiling table.")
print("The original dataframe remains unchanged for the next stage.")


INITIAL NUMERICAL PROFILE


Number of numerical columns identified: 12

Numerical columns:
1. amount_INR
2. Fraud Flag
3. hour of day
4. is weekend
5. battery level %
6. device storage free gb
7. screen brightness %
8. GPS accuracy m
9. nearby wifi count
10. session duration sec
11. notifications last hour
12. merchant distance km

Basic numerical summary:


,count,mean,std,min,25%,50%,75%,max
amount_INR,230681.0,1288.805363,1829.882227,10.0,284.0,615.00,1554.00,42099.0
Fraud Flag,236088.0,0.001902,0.043569,0.0,0.0,0.00,0.00,1.0
hour of day,242976.0,15.679664,10.931996,-1.0,11.0,16.00,19.00,99.0
is weekend,239795.0,0.290411,0.506763,-1.0,0.0,0.00,1.00,2.0
battery level %,240510.0,52.986836,30.496110,-10.0,28.0,52.00,77.00,150.0
device storage free gb,241200.0,1642.065501,12207.473086,-5.0,63.7,129.10,194.50,99999.0
screen brightness %,241175.0,55.408322,29.430525,-10.0,32.0,55.00,78.00,150.0
GPS accuracy m,241272.0,1543.373568,12261.654160,-5.0,8.6,13.50,21.20,99999.0
nearby wifi count,241358.0,19.322591,122.841771,-1.0,3.0,4.00,5.00,999.0
session duration sec,241242.0,14812.520498,120468.769729,-20.0,41.0,71.00,113.00,999999.0



Numerical profiling note:
Numeric-looking fields were converted only in a temporary profiling table.
The original dataframe remains unchanged for the next stage.


# Stage 1 Conclusion

The UPI transaction dataset was successfully loaded and inspected using the
same Stage 1 workflow used for the project.

The initial inspection established the dataset dimensions, complete column
structure, data types, record samples, unique-value information, missing-value
status, duplicate-row status and numerical-field profile.

The current dataset contains **250,000 records across 27 columns**.
The initial data-quality inspection identified **124,747 missing cells** and
**0 duplicate rows**.

Several numerical fields are stored as object/text values in the raw CSV. For
Stage 1, these fields were temporarily converted only for numerical profiling;
the original dataframe was not modified.

These results establish the baseline structure and quality of the raw dataset.
The dataset can therefore proceed to the next stage, where the relevant records
and fields will be acquired and filtered according to the project requirements.
